File: example/src/main/java/com/github/vivchar/example/widgets/MyItemDecoration.java
package com.github.vivchar.example.widgets;

import android.content.res.Resources;
import android.graphics.Rect;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import com.github.vivchar.example.pages.github.items.CategoryModel;
import com.github.vivchar.example.pages.github.items.RecyclerViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;

/**
 * Created by Vivchar Vitaly on 7/24/17.
 */

public class MyItemDecoration extends RecyclerView.ItemDecoration {

	@Override
	public void getItemOffsets(final Rect outRect, final View view, final RecyclerView parent, final RecyclerView.State state) {
		final RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();

		final int itemPosition = parent.getChildAdapterPosition(view);
		if (itemPosition != RecyclerView.NO_POSITION) {
			if (layoutManager instanceof GridLayoutManager) {
				final RendererRecyclerViewAdapter adapter = (RendererRecyclerViewAdapter) parent.getAdapter();
				final ViewModel item = adapter.getItem(itemPosition);
				if (item instanceof RecyclerViewModel) {
					outRect.left = dpToPixels(-10);
					outRect.right = dpToPixels(-10);
					outRect.top = dpToPixels(5);
					outRect.bottom = dpToPixels(5);
				} else if (item instanceof CategoryModel) {
					outRect.left = dpToPixels(5);
					outRect.right = dpToPixels(5);
					outRect.top = dpToPixels(10);
					outRect.bottom = dpToPixels(2);
				} else {
					outRect.left = dpToPixels(5);
					outRect.right = dpToPixels(5);
					outRect.top = dpToPixels(5);
					outRect.bottom = dpToPixels(5);
				}
			} else {
				throw new UnsupportedClassVersionError("Unsupported LayoutManager");
			}
		}
	}

	private static int dpToPixels(final float dp) {
		return (int) (dp * Resources.getSystem().getDisplayMetrics().density);
	}
}

File: example/src/main/java/com/github/vivchar/example/widgets/MyItemDecoration.kt
package com.github.vivchar.example.widgets

import android.graphics.Rect
import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.github.vivchar.example.pages.github.items.CategoryModel
import com.github.vivchar.example.pages.github.items.RecyclerViewModel
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel

/**
 * Created by Vivchar Vitaly on 7/24/17.
 */
class MyItemDecoration: ItemDecoration() {

	override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
		val layoutManager = parent.layoutManager
		val itemPosition = parent.getChildAdapterPosition(view)
		if (itemPosition != RecyclerView.NO_POSITION) {
			if (layoutManager is GridLayoutManager) {
				val adapter = parent.adapter as RendererRecyclerViewAdapter?
				when (adapter!!.getItem<ViewModel>(itemPosition)) {
					is RecyclerViewModel -> {
						outRect.left = (-10f).dpToPixels()
						outRect.right = (-10f).dpToPixels()
						outRect.top = (5f).dpToPixels()
						outRect.bottom = (5f).dpToPixels()
					}
					is CategoryModel -> {
						outRect.left = (5f).dpToPixels()
						outRect.right = (5f).dpToPixels()
						outRect.top = (10f).dpToPixels()
						outRect.bottom = (2f).dpToPixels()
					}
					else -> {
						outRect.left = (5f).dpToPixels()
						outRect.right = (5f).dpToPixels()
						outRect.top = (5f).dpToPixels()
						outRect.bottom = (5f).dpToPixels()
					}
				}
			} else {
				throw UnsupportedClassVersionError("Unsupported LayoutManager")
			}
		}
	}
}